### PR TITLE
Made asyncio TaskGroup work with eager task factories

### LINKED ISSUE #764: `get_coro` is missing a `None` check in `_task_started`
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

4.4.0

### Python version

3.12

### What happened?

Hello!

This was discovered upstream in CPython, see [this issue](https://github.com/python/cpython/issues/122332). In AnyIO's `_task_started` function (see [here](https://github.com/agronholm/anyio/blob/master/src/anyio/_backends/_asyncio.py#L335)), `get_coro()` can return `None` as of 3.12, with asyncio's new `eager_task_factory`. 

Unfortunately, there's not all that much you can do about this at this very moment -- a bug in C caused this code to segfault on current versions (this is [fixed now](https://github.com/python/cpython/pull/122338), but not released).

### How can we reproduce the bug?

This was the reproducer used in the CPython issue, which eventually fails at `_task_started`. It's not all that minimal, sorry! I haven't used AnyIO enough to come up with my own reproducer.

```py
import asyncio

import httpx
import uvicorn
from fastapi import FastAPI
from starlette.responses import StreamingResponse


async def main():
    loop = asyncio.get_running_loop()
    loop.set_task_factory(asyncio.eager_task_factory)

    app = FastAPI()

    @app.get("/")
    async def get():
        async def _():
            yield "1"

        return StreamingResponse(
            _(),
        )

    server = uvicorn.Server(
        uvicorn.Config(
            app,
            host="0.0.0.0",
            port=8080,
        )
    )
    asyncio.create_task(server.serve())

    client = httpx.AsyncClient()
    await client.get("http://localhost:8080")


if __name__ == "__main__":
    asyncio.run(main())
```

Again, this code segfaults on 3.12.4, you'll need to wait for the next patch release for the fix. In the meantime, it might be a good idea to throw an exception if using AnyIO with `eager_task_factory`.

### PR DESCRIPTION
**Pull request by agronholm**
## Changes

Fixes #764.

This updates the task group code and type annotations to accommodate eager task factories.

Additionally, this updates pre-commit modules. The updated `mypy` catches the annotation issues due to its updated `typeshed`.

## Checklist

If this is a user-facing code change, like a bugfix or a new feature, please ensure that
you've fulfilled the following conditions (where applicable):

- [x] You've added tests (in `tests/`) added which would fail without your patch
- [ ] You've updated the documentation (in `docs/`, in case of behavior changes or new
features)
- [X] You've added a new changelog entry (in `docs/versionhistory.rst`).

If this is a trivial change, like a typo fix or a code reformatting, then you can ignore
these instructions.

### Updating the changelog

If there are no entries after the last release, use `**UNRELEASED**` as the version.
If, say, your patch fixes issue #123, the entry should look like this:

`* Fix big bad boo-boo in task groups (#123
<https://github.com/agronholm/anyio/issues/123>_; PR by @yourgithubaccount)`

If there's no issue linked, just link to your pull request instead by updating the
changelog after you've created the PR.

### CODE REVIEW

File: src/anyio/_backends/_asyncio.py [hunk 1]
```diff
@@ -844,16 +888,22 @@ def task_done(_task: asyncio.Task) -> None:
...
-
         # Make the spawned task inherit the task group's cancel scope
-        _task_states[task] = TaskState(
+        _task_states[coro] = task_state = TaskState(
             parent_id=parent_id, cancel_scope=self.cancel_scope
         )
+        name = get_callable_name(func) if name is None else str(name)
+        try:
+            task = create_task(coro, name=name)
+        except BaseException:
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:898]: This should just be a finally

**Comment by agronholm** [comment]
Sigh...I improved my test and it's failing now. Back to the drawing board.

**Comment by graingert** [comment]
> Sigh...I improved my test and it's failing now. Back to the drawing board.

CI looks green to me, what's failing?

**Comment by agronholm** [comment]
> > Sigh...I improved my test and it's failing now. Back to the drawing board.
> 
> CI looks green to me, what's failing?

Fetching task state didn't work, as my amended test showed. I've pushed it now.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -22,14 +22,14 @@ repos:
       - id: trailing-whitespace
 
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: v0.6.9
+    rev: v0.7.1
```
Comment (graingert-coef) [inline .pre-commit-config.yaml:25]: I think there's 0.8 out now?

Comment (graingert) [reply .pre-commit-config.yaml:25]: I think there's 0.8 out now?

File: docs/versionhistory.rst [hunk 3]
```diff
@@ -5,6 +5,12 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 
 **UNRELEASED**
 
+- Improved compatibility with asyncio's eager task factories:
+
+  * Updated the annotation of ``TaskInfo.coro`` to allow it to be ``None``
```
Comment (graingert) [inline docs/versionhistory.rst:10]: Is it possible for TaskInfo.coro to be None? Can't you only access it from a running task?

Comment (agronholm) [reply docs/versionhistory.rst:10]: This stemmed from a `mypy` complaint, but I think you're right – `TaskInfo` objects should never be created from completed tasks under any circumstances.

File: docs/versionhistory.rst [hunk 4]
```diff
@@ -5,6 +5,8 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 
 **UNRELEASED**
 
+- Updated ``TaskGroup`` to work with asyncio's eager task factories
```
[the remark this answers is not in the document]
Comment (agronholm) [reply docs/versionhistory.rst:8]: I'm not very inclined to apply this wording. I at least wouldn't know what that second sentence means, and I think it's too early to say anything about the future of eager task factories.

[the remark this answers is not in the document]
Comment (graingert) [reply docs/versionhistory.rst:8]: I think we should at least say experimental until the test suite is passing

[the remark this answers is not in the document]
Comment (agronholm) [reply docs/versionhistory.rst:8]: The `TaskGroup` test for this is passing though, and it's the only change we claim was made. Perhaps we could add a note in the task group's actual docs about this?

[the remark this answers is not in the document]
Comment (graingert) [reply docs/versionhistory.rst:8]: Yeah something to say it's experimental and different from trio behavior. If we find out it's too hard to support both behaviors from user code we can add a workaround without making a major release

**Review by graingert — APPROVED** [review]

**Pull request merged**